### PR TITLE
feat(types): add Fields trait and derive macro

### LINKED ISSUE #13851: Add a new derive macro `#[derive(Fields)]`
### Background

Sometimes, we may want to construct a `Row` or `DataChunk` with type safety. To achieve that, we can create a struct, and then do one of the following thing:

1. convert it to an `OwnedRow`
2. impl `Row` for the struct directly (not covered by the macro now)
3. Encode it to a chunk (not covered by the macro now)

There is one fundamental prerequisite, which is to know the data_types or schema corresponding to this struct.

The macro can help to do that.

Related https://github.com/risingwavelabs/risingwave/issues/13433

### Design

First of all, we'll introduce a new trait `Fields` in common crate.

```rust
pub trait Fields {
    pub fn fields() -> Vec<Field>;
}
```

Then we'll introduce a derive macro to help struct to implement it:

```rust
#[derive(Fields)]
struct PgShadowRow {
  usename: String,
  usesysid: i32,
  usecreatedb: bool,
  usesuper: bool,
  userepl: bool,
  usebypassrls: bool,
  passwd: String,
  valuntil: Timestamptz,
  useconfig: Vec<String>,
}
```

Here is the type mapping table:

|rust type|rw type|
|---|---|
|bool|Boolean|
|i16|Int16|
|i32|Int32|
|i64|Int64|
|f32/F32|Float32|
|f64/F64|Float64|
|&'static str/String|Varchar|
|Date|Date|
|Time|Time|
|Timestamp|Timestamp|
|Timestamptz|Timestamptz|
|Interval|Interval|
|Vec<u8>/Bytes|Bytea|
|Jsonb|Jsonb|
|Int256|Int256|
|Decimal|Decimal|
|Vec<T>|DataType::List(mapping(T))|
|Option<T>|mapping(T)|
|`T impl Fields`| Struct(T::fields)| 

### Future Optimizations

We can support more field level options such as `alias`, `unnamed`, `ignore` or `repr` in `field` attribute, as needed.

### Discussions

_No response_

### Q&A

_No response_

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Resolve #13851 

https://github.com/risingwavelabs/risingwave/blob/585f6379c65649f92c4f71dfe5747714d6429ee3/src/common/src/types/fields.rs#L38-L76

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,6 +7,7 @@ members = [
   "src/cmd_all",
   "src/common",
   "src/common/common_service",
+  "src/common/fields-derive",
```
Comment (xxchan) [inline Cargo.toml:10]: prefer underscore, to be consistent with others

File: src/common/src/types/fields.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            v7: Vec<u8>,
+            v8: std::vec::Vec<i16>,
+            v9: Option<Vec<i64>>,
+            v10: std::option::Option<Vec<Option<f32>>>,
+            v11: Box<Timestamp>,
+            v14: Sub,
+        }
+
+        assert_eq!(
+            Data::fields(),
```
Comment (xxchan) [inline src/common/src/types/fields.rs:55]: Just a hint. You can use expect_test::Expect::assert_debug_eq so you don’t have to manually write the RHS 🤪

File: src/common/src/types/with_data_type.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+use std::sync::Arc;
+
+use bytes::Bytes;
+
+use super::{
+    DataType, Date, Decimal, Fields, Int256, Interval, JsonbRef, JsonbVal, StructType, Time,
+    Timestamp, Timestamptz, F32, F64,
+};
+
+pub trait WithDataType {
```
Comment (xxchan) [inline src/common/src/types/with_data_type.rs:25]: This is a little hard to understand why it’s needed. Add some comments?

**Review by xxchan — COMMENTED** [review]
Others LGTM

File: src/common/src/types/fields.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use super::DataType;
+
+/// A struct can implements `Fields` when if can be represented as a relational Row.
+pub trait Fields {
```
Comment (xxchan) [inline src/common/src/types/fields.rs:17]:
```suggestion
/// A struct can implements `Fields` when if can be represented as a relational Row.
///
/// Can be automatically derived by `#[derive(risingwave_common::Fields)]`.
pub trait Fields {
```

File: src/common/fields-derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::Data;
+use syn::DeriveInput;
+use syn::Field;
+use syn::Result;
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/common/fields-derive/src/lib.rs:20]: `cargo fmt` works on my machine 🤡

[the remark this answers is not in the document]
Comment (xxchan) [reply src/common/fields-derive/src/lib.rs:20]: Maybe you are not in the root dir?

File: Cargo.toml (hunk 1 above)
Comment (TennyZhuang) [reply Cargo.toml:10]: It’s already inconsistent now, although most of them are underscores, there are still several crates using dash.

File: src/common/src/types/with_data_type.rs (hunk 3 above)
Comment (TennyZhuang) [reply src/common/src/types/with_data_type.rs:25]: I can’t find similar thing that can be used, but I guess we may use it in somewhere. I’ll review it later.

Currently it’s just a macro helper.

File: src/common/src/types/with_data_type.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    }
+}
+
+impl_with_data_type!(String, DataType::Varchar);
+impl_with_data_type!(Date, DataType::Date);
+impl_with_data_type!(Time, DataType::Time);
+impl_with_data_type!(Timestamp, DataType::Timestamp);
+impl_with_data_type!(Timestamptz, DataType::Timestamptz);
+impl_with_data_type!(Interval, DataType::Interval);
+impl_with_data_type!(Vec<u8>, DataType::Bytea);
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/common/src/types/with_data_type.rs:102]: We may need both lines?

File: src/common/fields-derive/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        .into_iter()
+        .map(|field_rs| {
+            let Field {
+                // We can support #[field(ignore)] or other useful attributes here.
+                attrs: _attrs,
+                ident: name,
+                ty,
+                ..
+            } = field_rs;
+            let name = name.map_or("".to_string(), |name| name.to_string());
```
Comment (wangrunji0408) [inline src/common/fields-derive/src/lib.rs:62]: Why a struct field may have no name? 🤔

File: src/common/src/types/fields.rs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    /// When the struct being converted to an [`Row`](crate::row::Row) or a [`DataChunk`](crate::array::DataChunk), it schema must be consistent with the `fields` call.
+    fn fields() -> Vec<(&'static str, DataType)>;
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/common/src/types/fields.rs:19]: +1 for making the return value static. Maybe use lazy-static?

**Review by wangrunji0408 — APPROVED** [review]
Rest LGTM!

File: src/common/fields-derive/src/lib.rs (hunk 7 above)
Comment (xxchan) [reply src/common/fields-derive/src/lib.rs:62]: tuple

Comment (xiangjinwu) [reply src/common/fields-derive/src/lib.rs:62]: In PostgreSQL the names would implicitly become `f1`, `f2`, ...
```
test=# select row(9, 18);
  row   
--------
 (9,18)
(1 row)

test=# select (row(9, 18)).f1;
 f1 
----
  9
(1 row)

test=# select to_jsonb(row(9, 18));
      to_jsonb       
---------------------
 {"f1": 9, "f2": 18}
(1 row)
```

But in RisingWave this has not been supported consistently yet. Some places are empty while some places have these implicit names. Have not thought about the impact on backward compatibility if it was fixed to be pg-compatible.

No opinion on what the behavior of this place should be.

File: src/common/src/types/fields.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/common/src/types/fields.rs:19]: Got it, but it increases the complexity. We can do it when it’s really used in critical path.

File: src/common/fields-derive/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/common/fields-derive/src/lib.rs:20]: `brew uninstall rustfmt`

File: src/common/src/types/fields.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/common/src/types/fields.rs:21]: We can use an attribute `static` to control whether make it lazy static.

Please note that `static` doesn't mean always more efficient.
If user only called once, and need the ownership, then we need to allocate once, return a static reference, and then allocate (clone) again.

And we can modify the signature to this:

```rust
fn fields() -> impl ToOwned<Owned = Vec<(&'static str, DataType)>> + AsRef<[(&'static str, DataType)]>;
```

**Pull request merged**